Anonymous Apex Parser

ethan-sargent opened the pull request:
Created a new parser based on the existing apex parser, which supports parsing anonymous apex. 
A separate parser was chosen as anonymous apex allows any statement or declaration at a top level (excluding trigger declarations, but including method declarations instead). 
This conflicts with the top level statements acceptable in ordinary apex classes. 

Existing anonymous apex tests implemented in corpus.
Most existing corpus files minus trigger corpus files from apex parser retained.

ethan-sargent commented:
@aheber let me know if you would like me to squash these commits.

aheber commented:
@ethan-sargent, thank you for putting this together. I appreciate that looking through your changes you were able to make everything fit well.

I'm still unsure if it makes more sense to split anon apex into a separate parser language or leave it as a single.

Honestly, the only reason I lean into making it a separate language is because the Apex language server for VS Code (Jorje) sees it as a different language context, though that is largely because it's burden for spotting errors is different than this parse would be.

I guess I'm still unsure about the philosophical differences in Apex and Anonymous Apex. As I have worked on this library I started by giving it deep context about what is correct and not correct (which annotations, which SOQL functions, etc...) and I slowly removed some of that awareness because I decided that should be enforced at a higher level. Parse-ability != correctness, mostly when thinking along the lines of a language specification vs. the implementation.

Not to say I've removed all correctness checking but as I had to solve problems or was faced with my own complexity I often choose dropping correctness check in favor of a simpler definition that would survive expansion of the language better.

I'm not say yes or no to splitting the language, but I'd like to explain my reasoning before asking the question, that I'd like you to help me answer, should Anon Apex be separate, or should we just expand the supported parse structures for Apex?

Happy to have a long drawn out conversations about it, or just some quick thoughts then I can ruminate on how best to proceed.

I've been thinking about it for a couple of months and I'm leaning on not separating them, I honestly can't come up with any good reasons to separate them aside from contextual correctness. Anon Apex does have some interesting structures that would have to be added and I don't know if that will cause other cascading parse problems down the road but I expect we'll have those in the Anon Apex parsing no matter what we do.

One of the reasons for this is that after completing most of the parsing logic and running it against the various open source code bases I found that many of those projects stored Anon Apex as `.cls` files and broke the parsing logic anyways, rather than the convention of using `.apex` for Anon Apex. I'm not saying we should wrap a tool's implementation around a few projects seemingly bad decisions, honestly I think they just pre-date convention and problem did that so they could get Java's syntax support (totally just a guess)

Sorry for my little novel, and my delay in replying. Appreciate your engagement and look forward to finalizing these changes from you and rounding out this corner of the implementation.

ethan-sargent commented:
@aheber Thanks for reaching back out!

I think you're right on unifying the parsers. 
I've actually been working on a branch (feature/unify-apex) that does this, both for ease of maintenance and for the same reasons you've mentioned above of focusing on parse-ability over language correctness. I agree that it's not the responsibility of the parser to refuse to parse valid Apex constructs in incorrect locations (eg. method declarations outside of a class).

I noticed as well that you had some context-awareness integrated into the grammars which I was considering removing - the dialect-based parsing is interesting but probably not a good separation of concerns.

Philosophically I think the grammar should treat both anonymous apex and apex files as a series of statements.
I've been daily driving the Apex parser on both Anon Apex and Apex by modifying the parser output from a choice of declarations to a series of statements, and it works wonderfully to parse both files.

If you're on board with this, let me know, and I'll raise a PR with the updated Apex parser.

I've also added in support for the optional chaining operator + highlights, created an indents.scm file, and fixed a bug where a unary operator highlight applied to the whole expression following it rather than just to the unary operator. I can raise PRs with those separately, they are currently in my dotfiles repo rather than the parser repo.
